Services/communication_service.c

```diff
@@ -37,155 +38,205 @@ SAT_returnState communication_service_app(csp_packet_t *packet) {
+        if (sizeof(S_TR) + 1 > csp_buffer_data_size()) {
+          return SATR_ERROR;
+        }
+        S_TR.transmit = csp_hton32((uint32_t)S_TR.transmit);
         // what to do for int?
-        status = 0;
         memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
-        memcpy(&packet->data[OUT_DATA_BYTE], &S_buffer.transmit, sizeof(int));
-        set_packet_length(packet, sizeof(int8_t) + sizeof(int) + 1);
+        memcpy(&packet->data[OUT_DATA_BYTE], &S_TR, sizeof(S_TR));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(S_TR) + 1);
         if (queue_response(packet) != SATR_OK) {
           return SATR_ERROR;
         }
         break;
 
+    case GET_HK:
+        status = HAL_S_getHK (&S_HK);
+        if (sizeof(S_HK) + 1 > csp_buffer_data_size()) {
+          return SATR_ERROR;
+        }
+        /*Use va_list from stdarg for various number of arguments. */
+        S_HK.Output_Power = csp_htonflt(S_HK.Output_Power);
+        S_HK.PA_Temp = csp_htonflt(S_HK.PA_Temp);
+        S_HK.Top_Temp = csp_htonflt(S_HK.Top_Temp);
+        S_HK.Bottom_Temp = csp_htonflt(S_HK.Bottom_Temp);
+        S_HK.Bat_Current = csp_htonflt(S_HK.Bat_Current);
+        S_HK.Bat_Voltage = csp_htonflt(S_HK.Bat_Voltage);
+        S_HK.PA_Current = csp_htonflt(S_HK.PA_Current);
+        S_HK.PA_Voltage = csp_htonflt(S_HK.PA_Voltage);
+
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        memcpy(&packet->data[OUT_DATA_BYTE], &S_HK, sizeof(S_HK));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(S_HK) + 1);
+
+        if (queue_response(packet) != SATR_OK) {
+          return SATR_ERROR;
+        }
+              break;
+
+    case GET_BUFFER:
+      SID = packet->data[SID_byte];
+      status = HAL_S_getBuffer(SID, &S_buffer);
+      if (sizeof(S_buffer) + 1 > csp_buffer_data_size()) {
+        return SATR_ERROR;
+      }
+      S_buffer.pointer[SID] = csp_hton16(S_buffer.pointer[SID]);
+      memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+      memcpy(&packet->data[OUT_DATA_BYTE], &S_buffer.pointer[SID], sizeof(S_buffer.pointer[SID]));
+      set_packet_length(packet, sizeof(int8_t) + sizeof(S_buffer.pointer[SID]) + 1);
+      if (queue_response(packet) != SATR_OK) {
+        return SATR_ERROR;
+      }
+      break;
 
-/*    case SET_FREQ:
-      cnv8_32(&packet->data[IN_DATA_BYTE],&S_config.S_freq); //check what happens if skipped
-      S_config.S_freq = csp_ntoh32(S_config.S_freq);
+    case SOFT_RESET:
+      status = HAL_S_softResetFPGA();
+      memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+      set_packet_length(packet, sizeof(int8_t) + 1);
+      if (queue_response(packet) != SATR_OK) {
+        return SATR_ERROR;
+      }
+      break;
 
-      if (!WRITEDATA_ISOK(S_config.S_freq, ser_subtype)){
-          status = -1;
-      } else {
-          HAL_S_setFreq(S_config.S_freq);
-          status = 0;
+    case GET_FULL_STATUS:
```

Comment:
GS output:

to send: obc.communication.get_full_status
/usr/lib/python3.6/re.py:212: FutureWarning: split() requires a non-empty pattern match.
  return _compile(pattern, flags).split(string, maxsplit)
here
{'dst': 0, 'dport': 10, 'subservice': 10, 'args': bytearray(b'\n')}
CMP ident: (0, 'hostname', 'model', 'resvision', 'Oct 13 2020', '17:25:17')
Ping: 49 mS
bytearray(b'\n')
SENDING THE PACKET

0
10
WAIT FOR CONNECTION ... (CTRL+C to stop)
connection: source=0:49, dest=16:10
50
bytearray(b'\n\x00\x01\x01\x00\x00\x00\x01\x00\x00\x00\x01\x00\x02A\xd0\x00\x00A\xdaff\xc0333A;33B\xbe\x00\x00@\xe6ff>\xf5\xc2\x8f@\xa333@\xe4z\xe1')
{'err': 0, 'PWRGD': 1, 'TXL': 1, 'Transmit Ready': 1, 'Buffer Count': 0, 'Buffer Underrun': 1, 'Buffer Overrun': 2, 'Output Power': 26.0, 'Power Amplifier Temperature': 27.3, 'Top Temperature': -2.8, 'Bottom Temperature': 11.7, 'Battery Current': 95.0, 'Battery Voltage': 7.2, 'Power Amplifier Current': 0.48, 'Power Amplifier Voltage': 5.1, 'Firmware Version': 7.14}
packet is None; no more packets
WAIT FOR CONNECTION ... (CTRL+C to stop)
WAIT FOR CONNECTION ... (CTRL+C to stop)
^Cexit flag set to True (repeat to exit now)
Exiting receiving loop

Reply:
I'm keeping a log file of these TCs btw. Let me know if you want me to put it on a shared doc or something.

Reply:
I think this is good enough for record keeping, thanks! Sweet, looks like things are working like a charm :+1:

Reply:
Very satisfying.